src/main/scala/lms/core/types.scala

```diff
@@ -0,0 +1,129 @@
+package lms.core
+
+import lms.core.Backend.{Const, Exp, Sym}
+import lms.core.TypeMap.TypeExp
+
+import scala.collection.mutable
+
+
+object Reflect {
+  def unapply(x: Any)(implicit tm: TypeMap): Option[(String, List[Exp])] = x match {
+    case s@Sym(_) =>
+      tm.g.findDefinition(s).map(n => (n.op, n.rhs.filter(_.isInstanceOf[Exp]).map(_.asInstanceOf[Exp])))
+    case _ => None
+  }
+}
+
+abstract class Type[T] {
+  def reify(x: Exp, tm: TypeMap): TypeExp
+}
+
+abstract class SimpleType[T] extends Type[T] {
+  def reify(x: Exp, tm: TypeMap): TypeExp = reify(tm)
+  def reify(tm: TypeMap): TypeExp
```

Comment:
what does the `reify` function do? adding (key, value) pairs to `tm`?

Reply:
Yes, `reify` is called by `Wrap` to add key value pairs to `tm`